### PR TITLE
Guide, first draft: Configuring SPF and DKIM in Postfix on Debian 8

### PR DESCRIPTION
**Pull request by tknarr**
Configuring Postfix on Debian 8 (Jessie) to use SPF and DKIM to validate incoming email and add header signatures to outgoing email so other sites can validate it.

### CODE REVIEW

**Comment by tknarr** [comment]
The filenames for assets still have numbers prefixed, I'll fix that shortly.

**Comment by tknarr** [comment]
I'm thinking I ought to add in the instructions for adding backup MX servers to the configuration so SPF and DKIM checks are correctly bypassed for them. I don't see any guides that set up backup MXes, but I can't be the only one running one.

**Comment by tknarr** [comment]
Given the number of changes in the upstream, drop me a note if you'd like me to rebase this off the current master.

**Comment by alexfornuto** [comment]
Thanks @tknarr but we should be ok without a rebase. Sorry for the delay, but you're close to the top of the queue at this point.

File: docs/email/postfix/postfix-spf-dkim-debian-8.md [hunk 1]
```diff
@@ -0,0 +1,375 @@
...
+
+1. Allow mail from all hosts listed in the MX records for the domain:
+
+    "v=spf1 mx -all"
+
+2. Allow mail from a specific host:
+
+    "v=spf1 a:mail.example.com -all"
+
+The "v=spf1" tag is required and has to be the first tag. The last tag, "-all", indicates that mail from your domain should only come from servers identified in the SPF string, anything coming from any other source is forging your domain. An alternative is "~all", indicating the same thing but also indicating that mailservers should accept the message and flag it as forged instead of rejecting it outright. I prefer "-all" to make it harder for spammers to forge my domain successfully, but if you're nervous you might want to use "~all" so no mail gets lost because of errors.
```
Comment (alexfornuto) [inline docs/email/postfix/postfix-spf-dkim-debian-8.md]: The last sentence needs to be rewritten to remove the fist person, and make it less a personal opinion and more a recommendation.

File: docs/email/postfix/postfix-spf-dkim-debian-8.md [hunk 2]
```diff
@@ -0,0 +1,375 @@
...
+
+To add the record, go to your DNS management interface and add a record of type TXT for your domain itself (ie. a blank hostname) containing this string:
+
+    "v=spf1 mx -all"
+
+If you're using Linode's DNS Manager, go to the domain zone page for the domain you want to set SPF up for and add a new TXT record. The screen will look something like this when you've got it filled out:
+
+![Linode DNS manager add TXT record](/docs/assets/Postfix_SPF_TXT_record.png)
+
+If your DNS provider allows it (DNS Manager doesn't) you should also add a record of type SPF, filling it in the same way as you did the TXT record.
```
Comment (alexfornuto) [inline docs/email/postfix/postfix-spf-dkim-debian-8.md]: This is not quite right. You would add `spf` under the Name section.

File: docs/email/postfix/postfix-spf-dkim-debian-8.md [hunk 3]
```diff
@@ -0,0 +1,375 @@
...
+
+## Adding the SPF policy agent to Postfix ##
+
+The Python SPF policy agent adds SPF policy checking to Postfix. The SPF record for the sender's domain for incoming mail will be checked and if it exists mail will be handled accordingly. I selected the Python version because it's more capable and more up-to-date than the rather basic variant written in Perl.
+
+You'll need to switch to a root shell using `su -` or `sudo -s` for this.
```
Comment (alexfornuto) [inline docs/email/postfix/postfix-spf-dkim-debian-8.md]: By specifying the need for root access in the beginning of the guide, you've told the reader that all commands below need to be run as `root` or with `sudo`. There's no need to repeat it throughout the guide.

**Comment by alexfornuto** [comment]
I'm still reviewing this submission and making inline notes as a go, but here's PR to show some general formatting fixes that need to be made throughout: https://github.com/tknarr/linode-docs/pull/1

File: docs/email/postfix/postfix-spf-dkim-debian-8.md [hunk 4]
```diff
@@ -0,0 +1,375 @@
...
+
+1. If you are using SpamAssassin to do spam filtering, you may want to edit `/etc/postfix-policyd-spf-python/policyd-spf.conf` to change the `HELO_reject` and `Mail_From_reject` settings to `False`. That will cause the SPF policy agent to run it's tests and and add a message header with the results in it but _not_ reject any messages. You may also want to make this change if you want to see the results of the checks but not actual apply them to mail processing. Otherwise, just go with the standard settings.
+
+2. Edit `/etc/postfix/master.cf` and add the following entry at the end:
+
+    policyd-spf  unix  -       n       n       -       0       spawn
```
Comment (alexfornuto) [inline docs/email/postfix/postfix-spf-dkim-debian-8.md]: Please use the [file-excerpt](https://linode.com/docs/linode-writers-formatting-guide#files-and-file-excerpts) tag to show snippets that belong in a file. This will also help show the correct indentation as it should be in the file.

File: docs/email/postfix/postfix-spf-dkim-debian-8.md (hunk 2 above)
Comment (tknarr) [inline docs/email/postfix/postfix-spf-dkim-debian-8.md]: It would be right. SPF is a different RR type from TXT (99 instead of 16) and uses the same hostname as the TXT form. There's been a running dispute between the web and DNS factions within the working group, with the web faction favoring using TXT records for anything new (they exist and all DNS software handles them, it's just a matter of stating what the rules are for the content) and the DNS faction favoring defining a new RR for any purpose where the record content has a specific format (TXT's content is defined as free-form text, if your record requires a particular syntax as opposed to free-form text for the content you shouldn't be using TXT records for it except during prototyping/proof-of-concept). I favor the DNS faction on the grounds that just because you can use a hammer to put a screw in doesn't mean you shouldn't go get a screwdriver (or preferably a set), and that it makes it easier to syntax-check things.

**Comment by tknarr** [comment]
Pushed the cleanup based on comments above and https://github.com/tknarr/linode-docs/pull/1

**Comment by alexfornuto** [comment]
Thanks @tknarr I'll be taking another look at this first thing in the morning.

File: docs/email/postfix/postfix-spf-dkim-debian-8.md [hunk 5]
```diff
@@ -0,0 +1,376 @@
...
+
+    1.  Add this entry to increase the timeout Postfix uses with the policy agent to prevent Postfix from aborting the agent if transactions are running a bit slowly:
+
+{: .file-excerpt}
+/etc/postfix/main.cf
+:   ~~~ conf
+    policy_time_limit = 3600
+    ~~~
+
+    2.  Edit the `smtpd_recipient_restrictions` entry to add a `check_policy_service` entry thusly:
```
Comment (alexfornuto) [inline docs/email/postfix/postfix-spf-dkim-debian-8.md]: On my Debian 8 installation, this section does not yet exist in `main.cf`.

File: docs/email/postfix/postfix-spf-dkim-debian-8.md [hunk 6]
```diff
@@ -0,0 +1,376 @@
...
+        ...
+    ~~~
+
+    Make sure to add the `check_policy_service` entry **after** the `reject_unauth_destination` entry to avoid having your system become an open relay.
+
+4.  Restart Postfix with:
+
+    systemctl restart postfix
+
+You can check the operation of the policy agent by looking at raw headers on incoming mail messages for the SPF results header. You'll find errors logged in `/var/log/mail.log`.
```
Comment (alexfornuto) [inline docs/email/postfix/postfix-spf-dkim-debian-8.md]: I'm not seeing any differences in mail headers or in mail.log for messages sent from valid and invalid senders. Can you clarify what the reader should look for that's different?

File: docs/email/postfix/postfix-spf-dkim-debian-8.md [hunk 7]
```diff
@@ -0,0 +1,376 @@
...
+6.  Generate keys for each domain. First do a `cd /etc/opendkim/keys`. Then issue the following command:
+
+    opendkim-genkey -b 2048 -r -s YYYYMM
+
+replacing YYYYMM with the current year and month as in the key table. This will give you two files, `YYYYMM.private` containing the key and `YYYYMM.txt` containing the TXT record you'll need to set up DNS. Rename the files so they have names matching the third section of the second field of the key table for the domain:
+
+    mv YYYYMM.private example.private
+    mv YYYYMM.txt example.txt
+
+Repeat the commands in this step for every entry in the key table. The `-b 2048` indicates the number of bits in the RSA key pair used for signing and verification. 1024 bits is the minimum, but with modern hardware 2048 bits is safer and it's possible 4096 bits will be required at some point.
```
Comment (alexfornuto) [inline docs/email/postfix/postfix-spf-dkim-debian-8.md]: Is there any reason not to be using 4096 bit keys at this point?

File: docs/email/postfix/postfix-spf-dkim-debian-8.md [hunk 8]
```diff
@@ -0,0 +1,376 @@
...
+    myhostname
+    myhostname.example.com
+    example.com
+    ~~~
+
+When creating the file, change `myhostname` to the name of your server and replace `example.com` with your own domain name. We're identifying the hosts that users will be submitting mail through and that should have their outgoing mail signed, which for basic configurations will just be your own mail server.
+
+6.  Generate keys for each domain. First do a `cd /etc/opendkim/keys`. Then issue the following command:
+
+    opendkim-genkey -b 2048 -r -s YYYYMM
```
Comment (alexfornuto) [inline docs/email/postfix/postfix-spf-dkim-debian-8.md]: After running this command, I'm left with a text file that ends with:

```
DKIM key 201512 for example.com
```

I've double checked that all my configuration files from this guide use my domain and not `example.com`. Are there other configuration files not mentioned here, or does this always use that term?

**Comment by alexfornuto** [comment]
I don't know if it's because of the issue I noted on L217, but when I attempt to test my key I keep getting errors like this:

```
opendkim-testkey: syntax error in key data (ASCII 0x22 at offset 1)
```

**Comment by tknarr** [comment]
Thanks. I'll check these and make fixes or clarify things in the document. I should have an update waiting for you first thing in the morning.

File: docs/email/postfix/postfix-spf-dkim-debian-8.md (hunk 5 above)
Comment (tknarr) [inline docs/email/postfix/postfix-spf-dkim-debian-8.md]: The smtpd_recipient_restrictions entry? It won't if you haven't completed configuring Postfix for normal Internet operation without being an open relay (that's why the second note, I'm converting that to a caution to emphasize it).

File: docs/email/postfix/postfix-spf-dkim-debian-8.md (hunk 7 above)
Comment (tknarr) [inline docs/email/postfix/postfix-spf-dkim-debian-8.md]: RFC 6373 section 3.3.3 (http://www.ietf.org/rfc/rfc6376.txt). 2048 bits is the largest key verifiers are required to support, signatures from 4096-bit keys may not fit in a UDP DNS response packet and the verification starts to load up the mail server. The small key size is why key rotation's needed and why they recommend a monthly rotation.

File: docs/email/postfix/postfix-spf-dkim-debian-8.md (hunk 8 above)
Comment (tknarr) [inline docs/email/postfix/postfix-spf-dkim-debian-8.md]: It always uses example.com in the .txt file, which is just a snippet of a BIND zone file and gets used to set up DNS. The next step snips out just the record itself from between the parentheses and for Linode's DNS Manager you paste it into the form field as the record contents, the comment containing the domain name gets left behind.

The .private file with the private key should be a base64-encoded RSA key file. with the marker lines at the start and end.

**Comment by tknarr** [comment]
Added the clarifications, emphasis on the part of the DKIM public key that needs extracted from the file opendkim-genkey creates, and quoted some asterisks to keep them from being incorrectly interpreted as markup characters.

File: docs/email/postfix/postfix-spf-dkim-debian-8.md (hunk 5 above)
Comment (alexfornuto) [inline docs/email/postfix/postfix-spf-dkim-debian-8.md]: Yea, looks like something I missed from my own guide. The shame!

File: docs/email/postfix/postfix-spf-dkim-debian-8.md [hunk 9]
```diff
@@ -0,0 +1,380 @@
...
+
+    Make sure to add the `check_policy_service` entry **after** the `reject_unauth_destination` entry to avoid having your system become an open relay. If `reject_unauth_destination` is the last item in your restrictions list, add the comma after it and omit the comma at the end of the `check_policy_service` item above.
+
+4.  Restart Postfix with:
+
+    systemctl restart postfix
+
+You can check the operation of the policy agent by looking at raw headers on incoming mail messages for the SPF results header. You'll find errors logged in `/var/log/mail.log`. The header the policy agent adds should look something like this (asterisks are used to block identifying information):
+
+    Received-SPF: Pass (sender SPF authorized) identity=mailfrom; client-ip=***.*.*.**; helo=mail178-24.suw51.mandrillapp.com; envelope-from=********************@mandrillapp.com; receiver=******@***********.***
```
Comment (alexfornuto) [inline docs/email/postfix/postfix-spf-dkim-debian-8.md]: The messages in my `mail.log` that most closely resemble this are:

```
Dec 28 15:27:39 hostname policyd-spf[3927]: None; identity=helo; client-ip=2607:blah:blah:blah::; helo=mail-blah-blah.google.com; envelope-from=me(external account); receiver=me(internally)
Dec 28 15:27:39 hostname policyd-spf[3927]: Pass; identity=mailfrom; client-ip=2607::blah::blah; helo=mail-blah-blah.google.com; envelope-from=me(external account); receiver=me(internally)
```

The main differences is a) your example output is formatted more for human readability and b) for each test I get a `None` then a `Pass` entry. Should we be concerned about the differences?

File: docs/email/postfix/postfix-spf-dkim-debian-8.md (hunk 7 above)
Comment (alexfornuto) [inline docs/email/postfix/postfix-spf-dkim-debian-8.md]: :+1:

File: docs/email/postfix/postfix-spf-dkim-debian-8.md [hunk 10]
```diff
@@ -0,0 +1,380 @@
...
+    opendkim-genkey -b 2048 -r -s YYYYMM
+
+replacing YYYYMM with the current year and month as in the key table. This will give you two files, `YYYYMM.private` containing the key and `YYYYMM.txt` containing the TXT record you'll need to set up DNS. Rename the files so they have names matching the third section of the second field of the key table for the domain:
+
+    mv YYYYMM.private example.private
+    mv YYYYMM.txt example.txt
+
+Repeat the commands in this step for every entry in the key table. The `-b 2048` indicates the number of bits in the RSA key pair used for signing and verification. 1024 bits is the minimum, but with modern hardware 2048 bits is safer and it's possible 4096 bits will be required at some point.
+
+7.  Make sure the ownership and permissions on `/etc/opendkim` and it's contents are correct by running the following commands:
```
Comment (alexfornuto) [inline docs/email/postfix/postfix-spf-dkim-debian-8.md]: Am I mistaken in thinking that at this point all files in `/etc/opendkim` except for the contents of the `keys` directory are already set up with these permissions and ownership? If so, we need only apply this step to that directory.

File: docs/email/postfix/postfix-spf-dkim-debian-8.md [hunk 11]
```diff
@@ -0,0 +1,380 @@
...
+
+    cd /etc
```
Comment (alexfornuto) [inline docs/email/postfix/postfix-spf-dkim-debian-8.md]: My note above notwithstanding, since the commands below aren't relative, there's no reason to change the working directory is there?

File: docs/email/postfix/postfix-spf-dkim-debian-8.md [hunk 12]
```diff
@@ -0,0 +1,380 @@
...
+    "v=DKIM1; k=rsa; s=email; "
+    "p=MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAu5oIUrFDWZK7F4thFxpZa2or6jBEX3cSL6b2TJdPkO5iNn9vHNXhNX31nOefN8FksX94YbLJ8NHcFPbaZTW8R2HthYxRaCyqodxlLHibg8aHdfa+bxKeiI/xABRuAM0WG0JEDSyakMFqIO40ghj/h7DUc/4OXNdeQhrKDTlgf2bd+FjpJ3bNAFcMYa3Oeju33b2Tp+PdtqIwXR"
+    "ZksfuXh7m30kuyavp3Uaso145DRBaJZA55lNxmHWMgMjO+YjNeuR6j4oQqyGwzPaVcSdOG8Js2mXt+J3Hr+nNmJGxZUUW4Uw5ws08wT9opRgSpn+ThX2d1AgQePpGrWOamC3PdcwIDAQAB"
+    ~~~
+
+It's broken into chunks because of limitations in Bind (one of the most popular DNS server packages) and because of size limitations in the UDP protocol that's usually used for DNS requests and responses. Copy that region and paste it into the value for the TXT record.
+
+If you're using Linode's DNS manager, this is what the add TXT record screen will look like when you have it filled out:
+
+![Linode DNS manager add TXT record](/docs/assets/Postfix_DKIM_TXT_record.png)
```
Comment (alexfornuto) [inline docs/email/postfix/postfix-spf-dkim-debian-8.md]: The image shows a record name of `201510._domainkey` which isn't specified elsewhere in the doc. Is this required, common convention, or other?

File: docs/email/postfix/postfix-spf-dkim-debian-8.md [hunk 13]
```diff
@@ -0,0 +1,380 @@
...
+
+Repeat this for every domain you handle mail for, using the `.txt` file for that domain.
+
+### Test your configuration
+
+Test the keys for correct signing and verification using the `opendkim-testkey` command:
+
+    opendkim-testkey -d example.com -s YYYYMM
```
Comment (alexfornuto) [inline docs/email/postfix/postfix-spf-dkim-debian-8.md]: As noted earlier (though not in line) This test fails for me:

```
opendkim-testkey -d mydomain.com -s 201512 -vv
opendkim-testkey: using default configfile /etc/opendkim.conf
opendkim-testkey: checking key '201512._domainkey.mydomain.com'
opendkim-testkey: syntax error in key data (ASCII 0x22 at offset 0)
```

Looking at this full output alongside my note on L273, I'm wondering if it's related to the name given to the txt record.

**Comment by alexfornuto** [comment]
Here, have some more inline comments and another PR! https://github.com/tknarr/linode-docs/pull/2

This is really starting to shape up into a nice guide, I'm just not yet able to get past that DKIM key test.

File: docs/email/postfix/postfix-spf-dkim-debian-8.md (hunk 9 above)
Comment (tknarr) [inline docs/email/postfix/postfix-spf-dkim-debian-8.md]: Aha. In the log you're seeing the individual checks and their results. The header I gave was what'll appear in the message. I'll include an explanation of the log messages.

File: docs/email/postfix/postfix-spf-dkim-debian-8.md (hunk 10 above)
Comment (tknarr) [inline docs/email/postfix/postfix-spf-dkim-debian-8.md]: Permissions will probably be correct (read access is the only requirement), but the owner will be root for freshly-created files. So I fix ownership for the whole tree but just adjust the permissions for the keys directory and it's contents (I missed removing execute permission there, which should only matter for /etc/opendkim/keys itself) so only the owner can get at the files with the keys in them.

File: docs/email/postfix/postfix-spf-dkim-debian-8.md (hunk 12 above)
Comment (tknarr) [inline docs/email/postfix/postfix-spf-dkim-debian-8.md]: That's referenced up just under "### Setting up DNS", the names for the records will be of the form `YYYYMM._domainkeys` where YYYYMM is the year and month used in step 6 when generating the files. It was 201510 in my example because it was generated in October 2015 and I wanted the example to show what an actual selector would look like rather than a placeholder.

File: docs/email/postfix/postfix-spf-dkim-debian-8.md (hunk 13 above)
Comment (tknarr) [inline docs/email/postfix/postfix-spf-dkim-debian-8.md]: That should be correct, it should be 3 chunks each within double-quotes and separated from each other by whitespace. When I pull the TXT record for my domains that use DNS Manager, though, I get an extra set of quotes around the whole thing. The ones that use afraid.freedns.org for DNS are OK.
... augh. Does Linode's DNS manager automatically wrap whatever's in the value field in double-quotes? ... yes, yes it appears it does, the SPF records have the same double-quote problem. I'll have to fix that, and experiment to see how it handles multiple-chunk values too.

**Comment by tknarr** [comment]
I've got one TODO item about the log messages left, I'll fix that in a bit and push the change.

**Comment by tknarr** [comment]
Updated with the information about the mail.log messages added and the syntax of DNS TXT records updated to match DNS Manager. I also put in a note about possible differences between DNS Manager and other DNS providers' interfaces. That's what I had wrong with things, I'm used to freedns.afraid.org which takes TXT record values in exactly the format BIND's zonefiles use (quote marks and chunking), DNS Manager takes them as a single unquoted string and does the appropriate things with them.

**Comment by tknarr** [comment]
Updated the images for the SPF and DKIM examples so the values match what's written in the guide.

File: docs/email/postfix/postfix-spf-dkim-debian-8.md (hunk 13 above)
Comment (alexfornuto) [inline docs/email/postfix/postfix-spf-dkim-debian-8.md]: Success!

Comment (tknarr) [inline docs/email/postfix/postfix-spf-dkim-debian-8.md]: Good!

File: docs/email/postfix/postfix-spf-dkim-debian-8.md [hunk 14]
```diff
@@ -0,0 +1,401 @@
...
+2.  Edit `/etc/postfix/master.cf` and add the following entry at the end:
+
+    {: .file-excerpt}
+    /etc/postfix/master.cf
+    :   ~~~ text
+        policyd-spf  unix  -       n       n       -       0       spawn
+            user=policyd-spf argv=/usr/bin/policyd-spf
+        ~~~
+
+3.  Open `/etc/postfix/main.cf` and add this entry to increase the Postfix policy agent timeout, to prevent Postfix from aborting the agent if transactions are running a bit slowly:
```
Comment (alexfornuto) [inline docs/email/postfix/postfix-spf-dkim-debian-8.md]: `Jan 12 15:44:41 hostname postfix[4926]: /usr/sbin/postconf: warning: /etc/postfix/main.cf: unused parameter: policy_time_limit=3600`

Comment (tknarr) [inline docs/email/postfix/postfix-spf-dkim-debian-8.md]: Rrgh. That should've been policyd-spf_time_limit.

File: docs/email/postfix/postfix-spf-dkim-debian-8.md [hunk 15]
```diff
@@ -0,0 +1,401 @@
...
+
+    Make sure to add the `check_policy_service` entry **after** the `reject_unauth_destination` entry to avoid having your system become an open relay. If `reject_unauth_destination` is the last item in your restrictions list, add the comma after it and omit the comma at the end of the `check_policy_service` item above.
+
+5.  Restart Postfix with:
+
+        systemctl restart postfix
+
+You can check the operation of the policy agent by looking at raw headers on incoming mail messages for the SPF results header. You'll find errors logged in `/var/log/mail.log`. The header the policy agent adds to messages should look something like this:
+
+    Received-SPF: Pass (sender SPF authorized) identity=mailfrom; client-ip=127.0.0.1; helo=mail.example.com; envelope-from=text@example.com; receiver=tknarr@silverglass.org
```
Comment (alexfornuto) [inline docs/email/postfix/postfix-spf-dkim-debian-8.md]: The closest to this I have in my logs is:
```
policyd-spf[11571]: Pass; identity=mailfrom;
```

Is this another version of the same message?

**Comment by alexfornuto** [comment]
OK @tknarr We're almost there! I've completed my tech test (and applied it to my production email server, so thanks :+1: ).

 Since I made so many changes I'm no longer an impartial tester. @cwlinode is currently working on another bounty submission for an email server, so once that's set up he's going to run through this guide on top as an additional tech test. Once that's done we move on to copy edit, and then publishing! 

For now, just sit tight. @cwlinode will be in touch if he runs into any other issues.

**Comment by tknarr** [comment]
@alexfornuto On a related tangent, https://github.com/tknarr/opendkim-genkeys may help with generating/rotating DKIM keys.

**Comment by alexfornuto** [comment]
Reviewing internally, moving on to copy edit.

@tknarr that's an interesting project you've linked to. My first thought was that the README failed to provide a synopsis as to what the software does any how it works. It's a common mistake amongst OSS projects, developers assume that the context clues are enough to let a user know what's going on, but that isn't always the case.

**Comment by alexfornuto** [comment]
Our final edit is ready! https://github.com/tknarr/linode-docs/pull/3

**Comment by tknarr** [comment]
tknarr#3 looks good by me. Merged and ready.

**Comment by alexfornuto** [comment]
Great, please check your email for the final step.

**Comment by tknarr** [comment]
Got the email, sent the Paypal information. Thanks, and hope to hear back when it's processed.

**Comment by alexfornuto** [comment]
Thanks @tknarr I'm merging your guide now, it will be live in a few minutes. Your payment info has been sent to our accounting team, and you should see it within 24-48 hours. Thanks for writing a great guide, and working so well with us during the editing process.

**Comment by alexfornuto** [comment]
P.S. If you would like the guide to link to your github or twitter account, please let us know!

**Comment by tknarr** [comment]
Linking to my Github account would be ideal, thanks. https://github.com/tknarr

File: docs/email/postfix/configure-spf-and-dkim-in-postfix-on-debian-8.md [hunk 16]
```diff
@@ -0,0 +1,403 @@
...
+
+You don't need to set this up, but doing so makes it harder for anyone to forge email from your domains because recipient mail servers will see the lack of a DKIM signature and reject the message.
+
+### Key rotation
+
+The reason the YYYYMM format is used for the selector is that best practice calls for changing the DKIM signing keys every so often (monthly is recommended, and no longer than every 6 months). To do that without disrupting messages in transit, you generate the new keys using a new selector. The process is:
+
+1.  Generate new keys as in step 6 of "Configuring OpenDKIM". Do this in a scratch directory, not directly in `/etc/opendkim/keys`. Use the current year and month for the YYYYMM selector value, so it's different from the selector currently in use.
+
+2.  Use the newly-generated `.txt` files to add the new keys to DNS as in the DKIM "Setting Up DNS" section, using the new YYYYMM selector in the host names. Don't remove or alter the existing DKIM TXT records.
```
Comment (alexfornuto) [inline docs/email/postfix/configure-spf-and-dkim-in-postfix-on-debian-8.md]: I know it's a little late to think of this, but would it be a good idea at this point to instruct the reader to use the verify command before continuing to the next step and replacing the keys?

Comment (tknarr) [inline docs/email/postfix/configure-spf-and-dkim-in-postfix-on-debian-8.md]: Yes, that'd be a good idea. I should've thought of that too.

Comment (alexfornuto) [inline docs/email/postfix/configure-spf-and-dkim-in-postfix-on-debian-8.md]: I'll update it :wink:

Comment (tknarr) [inline docs/email/postfix/configure-spf-and-dkim-in-postfix-on-debian-8.md]: I made a slightly different change, it uses the -k option to opendkim-testkey to test things before moving the private key into place for Postfix. Pull request #392
